File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/replicatedentity/ReplicatedEntityProvider.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.replicatedentity;

import kalix.javasdk.Kalix;
import kalix.javasdk.impl.replicatedentity.ReplicatedEntityRouter;
import kalix.replicatedentity.ReplicatedData;
import com.google.protobuf.Descriptors;

/**
 * Register a value based entity in {@link Kalix} using a <code>
 *  ReplicatedEntityProvider</code>. The concrete <code>ReplicatedEntityProvider</code> is generated
 * for the specific entities defined in Protobuf.
 */
public interface ReplicatedEntityProvider<D extends ReplicatedData, E extends ReplicatedEntity<D>> {

  ReplicatedEntityOptions options();

  Descriptors.ServiceDescriptor serviceDescriptor();

  String typeId();

  ReplicatedEntityRouter<D, E> newRouter(ReplicatedEntityContext context);

  Descriptors.FileDescriptor[] additionalDescriptors();
}

File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/replicatedentity/ReplicatedEntityProvider.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk.replicatedentity

import kalix.replicatedentity.ReplicatedData
import kalix.scalasdk.impl.replicatedentity.ReplicatedEntityRouter
import com.google.protobuf.Descriptors

trait ReplicatedEntityProvider[D <: ReplicatedData, E <: ReplicatedEntity[D]] {

  def typeId: String
  def options: ReplicatedEntityOptions
  def newRouter(context: ReplicatedEntityContext): ReplicatedEntityRouter[D, E]

  def serviceDescriptor: Descriptors.ServiceDescriptor
  def additionalDescriptors: Seq[Descriptors.FileDescriptor]
}
